README

Implements the parser and interpreter of a non-turing-complete script language,
designed to model common data-acquisition, mockup and presentation-related
processing tasks.

Currently it covers the following aspects:
* Defining blocks of background processes, to be run once or periodically.
* Associate blocks of operations with a named "action", to be run by listeners.
* Expose actions through an HTTP (or other) interface (i.e. listeners).
* Deployment-specific configuration through well-defined variables.

Blocks of operations are lists of statements of which each statement operates on
a stack.

The example script "news" models the following application:
* Acquire news from various sources (via RSS / Atom feeds)
* Update those news feeds periodically
* Provide ability for a frontend application to perform the following actions:
  - Load all stores news entries, ordered by date ("load-all").
  - Load all stored news entries, which have not been read yet ("load-unread").
  - Flag a given news item as having been read ("flag-read").
  - Load a specific news item by key ("load")
  - Push a given news item onto a presumed "stack" for the user, being managed by
    the backend. Stacks are named to accomodate for multiple instances of an
    application, multiple users working at the same time and multiple contexts a
    single user may be working in.
  - Pop items from the stack, thus discarding them from current UI focus.

All code is written in `node --harmony` compatible ECMA script.


news

define "sources":
  type list
  required

define "news":
  type datastore
  required

listen:
  protocol http
  port 7845

task "update-news":
  # fetches all configured sources
  fetch sources
  # this will explode feed results into feed items ("entries")
  extract-feed-entries
  # this will extend each item with { "unread": true }
  set-flag "unread" true
  # this will store items in the abstract document store "news"
  persist news

background:
  # do whatever is in this block every hour. order within the block doesn't matter.
  every hour
  # run this task, as declared above.
  run-task "update-news"

# actions are invoked by the UI, either on behalf of the user or internally.
# they support both request / response and command / event semantics.
action "update":
  run-task "update-news"

action "load-all":
  # queries the abstract document store "news"
  load news
  # options for the query
  sort-by "date" desc

action "load":
  load-item news

action "load-unread":
  load news
  filter "unread" true
  sort-by "date" desc

action "load-read":
  load news
  filter "unread" false
  sort-by "date" desc

action "flag-read":
  load-item news
  set-flag "unread" false
  persist news


ui/news.js

var dom = {}, currentEntry = null, actions = {}, newsLoadAction;

window.addEventListener('load', function() {
  dom.list = document.getElementById('list');
  dom.content = document.getElementById('content');
  dom.controls = document.getElementById('controls');
  dom.list.nav = document.getElementById('list-nav');

  dom.list.nav.addEventListener('click', function(event) {
    if(event.target.tagName === 'A' && event.target.dataset.action) {
      setNewsLoadAction(event.target.dataset.action);
      loadNews();
    }
  });

  dom.controls.addEventListener('click', function(event) {
    if(event.target.tagName === 'BUTTON') {
      var action = event.target.dataset.action;
      actions[action]();
    }
  });

  setNewsLoadAction("load-unread");

  setInterval(updateNews, 20 * 60 * 1000);
  loadNews();
});

function setNewsLoadAction(action) {
  newsLoadAction = action;
  var current = dom.list.nav.getElementsByClassName('current');
  for(var i=0;i<current.length;i++) {
    current[i].classList.remove('current');
  }
  dom.list.nav.querySelector('*[data-action="' + action + '"]').classList.add('current');
}

actions['flag-read'] = function() {
  if(currentEntry) {
    runAction('flag-read', function() {
      var currentRow = getListItem(currentEntry._id);
      currentRow.nextElementSibling.click();
      currentRow.parentNode.removeChild(currentRow);
    }, currentEntry._id);
  }
};

function updateNews() {
  runAction("update", loadNews);
}

function loadNews() {
  setLoading(dom.list);
  runAction(newsLoadAction, renderNews);
}

function setLoading(domElement) {
  domElement.textContent = "Loading...";
}

function setCurrent(entry) {
  var current = dom.list.getElementsByClassName('current');
  for(var i=0;i<current.length;i++) {
    current[i].classList.remove('current');
  }
  currentEntry = entry;
  getListItem(entry._id).classList.add('current');
}

function getListItem(id) {
  var result = dom.list.querySelector('*[data-id="' + id + '"]');
  console.log('get item', id, '->', result);
  return result;
}

function renderNews(entries) {
  renderList(entries);
  if((! currentEntry) && entries.length > 0) {
    setCurrent(entries[0]);
    renderContent();
  } else {
    // doesn't re-render content, but highlights item in list.
    setCurrent(currentEntry);
  }
}

function renderContent() {
  if(currentEntry) {
    dom.content.innerHTML = "";
    var source = document.createElement('div');
    var key = document.createElement('strong');
    key.textContent = "Source: ";
    var value = document.createElement('a');
    value.textContent = currentEntry.feed.name;
    value.href = currentEntry.feed.link;
    source.appendChild(key);
    source.appendChild(value);
    dom.content.appendChild(source);
    var p = document.createElement('p');
    p.innerHTML = currentEntry.contentHTML;
    dom.content.appendChild(p);
  } else {
    dom.content.textContent = "No current entry.";
  }
}

function renderList(entries) {
  dom.list.innerHTML = '';
  if(entries.length === 0) {
    dom.list.textContent = "No entries.";
  } else {
    entries.forEach(function(entry) {
      dom.list.appendChild(renderEntry(entry));
    });
  }
}

function renderEntry(entry) {
  var li = document.createElement('li');
  li.dataset.id = entry._id;
  li.textContent = "(" + entry.feed.name + ") " + entry.title;
  li.addEventListener('click', function() {
    setCurrent(entry);
    renderContent();
  });
  return li;
}

function runAction(action, callback) {
  var args = Array.prototype.slice.call(arguments, 2);
  var path = [action].concat(args).join('/');
  var req = new XMLHttpRequest();
  req.open('POST', 'http://localhost:7845/' + path);
  req.addEventListener('load', function() {
    if(req.responseStatus === 500) {
      throw new Error("Failed to run action " + action + ": " + req.responseText);
    } else {
      var stack = JSON.parse(req.responseText);
      callback.apply(this, stack);
    }
  });
  req.send();
}


datastore.js
"use strict";

const events = require('events');
const levelup = require('levelup');

module.exports = function(path) {
  let db = levelup(path);
  let eventEmitter = new events.EventEmitter();

  return {
    on: eventEmitter.on.bind(eventEmitter),

    store: function(key, object, callback) {
      db.put(key, JSON.stringify(object), function(err) {
        if(err) {
          eventEmitter.emit('error', err);
        } else {
          callback(object);
        }
      });
    },

    load: function(callback) {
      let objects = [];
      db.createReadStream().
        on('data', function(item) {
          objects.push(JSON.parse(item.value));
        }).
        on('error', eventEmitter.emit.bind(eventEmitter, 'error')).
        on('end', function() {
          console.log('yielding', objects);
          callback(objects);
        });
    },

    loadItem: function(key, callback) {
      db.get(key, function(err, object) {
        if(err) {
          eventEmitter.emit('error', err);
        } else {
          callback(JSON.parse(object));
        }
      });
    }
  };
};


lib/feed-read.js
// Copied from https://github.com/sentientwaffle/feed-read/blob/master/index.js on Mar 24, 2014, 6 PM UTC.
// Original license: MIT style, https://github.com/sentientwaffle/feed-read/blob/master/LICENSE
// -- cutesquirrels

var request    = require('request')
  , sax        = require('sax')
  , _          = require('underscore');


// Public: Fetch the articles from the RSS or ATOM feed.
// 
// url      - The String feed url, or an Array of urls.
// callback - Receives `(err, articles)`, where each article has properties:
//          
//              * "title"
//              * "author"
//              * "link"
//              * "content"
//              * "published"
//              * "feed" - {name, source, link}
// 
// Returns nothing.
var FeedRead = module.exports = function(feed_url, callback) {
  if (feed_url instanceof Array) {
    var feed_urls = feed_url
      , articles  = [];
    var next = function(i) {
      var feed_url = feed_urls[i];
      if (!feed_url) return callback(null, articles);
      FeedRead.get(feed_url, function(err, _articles) {
        if (err) return callback(err);
        articles = articles.concat(_articles);
        next(i + 1);
      });
    };
    next(0);
  } else {
    FeedRead.get(feed_url, callback);
  }
};


// Public: Check if the XML is RSS, ATOM, or neither.
// 
// xml - A String of XML.
// 
// Returns "atom", "rss", or false when it is neither.
FeedRead.identify = function(xml) {
  if (/<rss /i.test(xml)) {
    return "rss";
  } else if (/<feed /i.test(xml)) {
    return "atom";
  } else {
    return false;
  }
}



// Internal: Get a single feed.
// 
// feed_url - String url.
// callback - Receives `(err, articles)`.
// 
FeedRead.get = function(feed_url, callback) {
  request(feed_url, function(err, res, body) {
    if (err) return callback(err);
    var type = FeedRead.identify(body);
    if (type == "atom") {
      FeedRead.atom(body, feed_url, callback);
    } else if (type == "rss") {
      FeedRead.rss(body, feed_url, callback);
    } else {
      return callback(new Error( "Body is not RSS or ATOM"
                                , body.substr(0, 30), "..."));
    }
  });
};



// Public: Parse the articles from some ATOM.
// 
// xml      - A XML String.
// source   - (optional)
// callback - Receives `(err, articles)`.
// 
// Returns an Array of Articles.
FeedRead.atom = function(xml, source, callback) {
  if (!callback) return FeedRead.atom(xml, "", source);
  
  var parser   = new FeedParser()
    , articles = []
    // Info about the feed itself, not an article.
    , meta     = {source: source}
    // The current article.
    , article
    // The author for when no author is specified for the post.
    , default_author;
  
  
  parser.onopentag = function(tag) {
    if (tag.name == "entry") article = tag;
  };
  
  parser.onclosetag = function(tagname, current_tag) {
    if (tagname == "entry") {
      articles.push(article);
      article = null;
    } else if (tagname == "author" && !article) {
      default_author = child_data(current_tag, "name");
    } else if (tagname == "link" && current_tag.attributes.rel != "self") {
      meta.link || (meta.link = current_tag.attributes.href);
    } else if (tagname == "title" && !current_tag.parent.parent) {
      meta.name = current_tag.children[0];
    } else if (tagname == "id" && !current_tag.parent.parent) {
      meta.guid = current_tag.children[0];
    }
  };
  
  parser.onend = function() {
    callback(null, _.map(articles,
      function(art) {
        var author = child_by_name(art, "author");
        if (author) author = child_data(author, "name");
        
        var obj = {
            title:     child_data(art, "title")
          , content:   child_data(art, "content")
          , published: child_data(art, "published")
                    || child_data(art, "updated")
          , author:    author || default_author
          , link:      child_by_name(art, "link").attributes.href
          , feed:      meta
          };
        if (obj.published) obj.published = new Date(obj.published);
        return obj;
      }
    ));
  };
  
  parser.write(xml);
};


// Public: Parse the articles from some RSS.
// 
// xml      - A XML String.
// source   - (optional)
// callback - Receives `(err, articles)`.
// 
// Returns an Array of Articles.
FeedRead.rss = function(xml, source, callback) {
  if (!callback) return FeedRead.rss(xml, "", source);
  
  var parser   = new FeedParser()
    , articles = []
    // Info about the feed itself, not an article.
    , meta     = {source: source}
    // The current article.
    , article;
  
  
  parser.onopentag = function(tag) {
    if (tag.name == "item") article = tag;
  };
  
  parser.onclosetag = function(tagname, current_tag) {
    if (tagname == "item") {
      articles.push(article);
      article = null;
    } else if (tagname == "channel") {
      meta.link || (meta.link = child_data(current_tag, "link"));
      meta.name = child_data(current_tag, "title");
    }
  };
  
  parser.onend = function() {
    callback(null, _.map(articles,
      function(art) {
        var obj = {
            title:     child_data(art, "title")
          , content:   scrub_html(child_data(art, "content:encoded"))
                    || scrub_html(child_data(art, "description"))
          , published: child_data(art, "pubDate")
          , author:    child_data(art, "author")
                    || child_data(art, "dc:creator")
          , link:      child_data(art, "link")
          , guid:      child_data(art, "guid")
          , feed:      meta
          };
        if (obj.published) obj.published = new Date(obj.published);
        return obj;
      }
    ));
  };
  
  parser.write(xml);
};


// Methods to override:
// 
//   * onopentag
//   * onclosetag
//   * onend
// 
var FeedParser = (function() {
  // Internal: Parse the XML.
  // 
  // xml      - An XML String.
  // callback - Receives `(err, obj)`.
  // 
  function FeedParser() {
    this.current_tag = null;
    var parser       = this.parser = sax.parser(true,
        { trim: true
        , normalize: true
        })
      , _this        = this;
    
    parser.onopentag  = function(tag) { _this.open(tag); };
    parser.onclosetag = function(tag) { _this.close(tag); };
    
    parser.ontext  = function(text) { _this.ontext(text); };
    parser.oncdata = function(text) { _this.ontext(text); };
    parser.onend   = function() { _this.onend(); };
    
    parser.onerror = console.error;
  }
  
  
  // Public: Parse the XML.
  FeedParser.prototype.write = function(xml) {
    this.parser.write(xml).close();
  };
  
  // Internal: Open a tag.
  FeedParser.prototype.open = function(tag) {
    tag.parent   = this.current_tag;
    tag.children = [];
    if (tag.parent) tag.parent.children.push(tag);
    this.current_tag = tag;
    this.onopentag(tag);
  };
  
  // Internal: CLose a tag.
  FeedParser.prototype.close = function(tagname) {
    this.onclosetag(tagname, this.current_tag);
    if (this.current_tag && this.current_tag.parent) {
      var p = this.current_tag.parent;
      delete this.current_tag.parent;
      this.current_tag = p;
    }
  };
  
  // Internal: Add the text as a child of the current tag.
  FeedParser.prototype.ontext = function(text) {
    if (this.current_tag) {
      this.current_tag.children.push(text);
    }
  };
  
  return FeedParser;
})();


// Internal: Remove <script> tags from the HTML.
// 
// html     - An HTML String.
// callback - Receives `(err, html)`.
// 
// TODO: Do actual HTML parsing!!
function scrub_html(html) {
  return html.replace(/<script.*<\/script>/gi, "");
}


// Internal: Find the first node from the parent node's children that has
// the given name.
// 
// parent - An Array of node objects.
// name   - String node name.
// 
// Returns a node Object or null.
function child_by_name(parent, name) {
  var children = parent.children || [];
  for (var i = 0; i < children.length; i++) {
    if (children[i].name == name) return children[i];
  }
  return null;
}

// Internal: Get the first child of `parent` with `name`,
// and return the text of its children.
function child_data(parent, name) {
  var node     = child_by_name(parent, name)
  if (!node) return "";
  var children = node.children;
  if (!children.length) return "";
  return children.join("");
}


parse-script.js
"use strict";
/**
 * This file is totally untidy. It contains, apart from the parser, the means to:
 * - convert a script's parse tokens into an AST
 * - resolve variables against builtin and configured symbols
 * - run script ASTs blocks in a stack-like environment
 * This file also contains a default configuration, bringing an in-memory key/val
 * store and an RSS feed fetcher.
 * Also an HTTP listener for "listen" blocks.
 * Also an implementation of every call issued in the "news" script.
 *
 * TODO: clean this up.
 */

const feedRead = require('./lib/feed-read');
const datastore = require('./datastore');

const fs = require('fs');
if(process.argv.length == 2) {
  console.log("Usage: " + process.argv.slice(0, 2).join(' ') + " <script>");
  process.exit(127);
}

const input = fs.createReadStream(process.argv[2]);

const BUILTINS = {
  day: { value: 24 * 3600 * 1000 }, // ms in a day
  hour: { value: 3600 * 1000 },
  minute: { value: 60 * 1000 },
  desc: { value: 'desc' },
  asc: { value: 'asc' }
};

const CORS_HEADERS = {
  'Access-Control-Allow-Origin': '*',
  'Access-Control-Allow-Methods': 'POST'
};

let buf = '';

input.on('data', function(chunk) {
  buf += chunk;
});

input.on('end', function() {

  /**
   * Type: context
   * Property: type
   * Property: name
   * Property: statements
   *
   * Type: block
   * Property: references (in occurrence order)
   * Property: calls (in statement order, may be nulled-out)
   * Property: timer (associated timer options)
   *
   * Type: reference
   * Property: type (value | call)
   * Property: symbol
   * Property: statement (index)
   * Property: arg (index)
   * (Property: call) (optional associated call)
   *
   * Type: call
   * Property: name (== reference.symbol)
   * Property: args
   *
   *
   */

  let contexts = [];
  let context = null;

  function pushContext(ctx) {
    if(context) {
      contexts.push(context);
    }
    context = ctx || null;
  }

  buf.split('\n').forEach(function(line, lineNo) {
    let md;
    line = line.split('#')[0];
    if(line.match(/^\s*$/)) {
      // empty, ignore.
    } else if((md = line.match(/^(task|background|action|define|listen)(.*)$/))) {
      let keyword = md[1], name = null;
      if(! md[2].match(/:$/)) {
        throw new Error("Expected block begin after keyword '" + keyword + "'!");
      }
      if((md = md[2].slice(0, -1).match(/\"(.+)\"/))) {
        name = md[1];
      }
      pushContext({ type: keyword, name: name, statements: [] });
    } else if(line.match(/^  (.+)$/)) {
      if(! context) {
        throw new Error("Statement without context in line " + (lineNo + 1) + "!");
      }
      let statement = line.slice(2);
      if(statement.match(/^\s/)) {
        throw new Error("Faulty indentation in line " + (lineNo + 1) + "!");
      }
      let tokens = [], tokenBuf = '';
      let string = false;
      function emitSymbol() {
        let token = {};
        if(tokenBuf.match(/^\d+$/)) {
          token.type = 'integer';
          token.value = parseInt(tokenBuf);
        } else if(tokenBuf.match(/^[\d\.]+$/)) {
          token.type = 'float';
          token.value = parseFloat(tokenBuf);
        } else if(tokenBuf.match(/^(?:true|false)$/)) {
          token.type = 'boolean';
          token.value = (tokenBuf === 'true');
        } else {
          token.type = 'symbol';
          token.value = tokenBuf;
        }
        tokens.push(token);
      }
      for(let i = 0, n = statement.length; i < n; i++) {
        let c = statement[i];
        if(c === '"') {
          if(string) {
            string = false;
            tokens.push({ type: 'string', value: tokenBuf });
            tokenBuf = '';
          } else if(tokenBuf.length > 0) {
            throw new Error("Unexpected string begin at " + (lineNo + 1) + ":" + (i + 2) + "!");
          } else {
            string = true;
          }
        } else if(c == ' ' && (! string)) {
          if(tokenBuf.length > 0) {
            emitSymbol();
            tokenBuf = '';
          }
        } else {
          tokenBuf += c;
        }
      }
      if(tokenBuf.length > 0) {
        emitSymbol();
      }
      context.statements.push(tokens);
    } else {
      throw new Error("Unmatched line " + lineNo);
    }
  });
  if(context) {
    pushContext();
  }

  console.log('parse result', JSON.stringify(contexts, null, 2));

  let script = evalScript(contexts);
  resolveSymbols(script);

  let sources = [
    {
      url: "http://blog.fefe.de/rss.xml?html",
      trustedHTML: true
    },
    {
      url: "http://www.tagesschau.de/xml/rss2"
    }
  ];

  let newsStore = datastore("db/news");
  newsStore.on('error', function(error) {
    console.log('newsStore failed: ' + error);
  });

  runScript(script, {
    sources: sources.map(function(source) {
      return {
        fetch: function(callback) {
          feedRead(source.url, function(err, entries) {
            let result = {
              source: source
            };
            if(err) {
              console.log("Failed to get entries from feed: " + err);
              result.entries = [];
            } else {
              result.entries = entries;
            }
            result.entries.forEach(function(entry) {
              if(source.trustedHTML) {
                entry.contentHTML = entry.content;
              } else {
                entry.contentHTML = entry.content.replace(/<[^>]+>/g, '');
              }
            });
            callback(result);
          });
        }
      };
    }),
    news: newsStore
  });
});

function runScript(script, config) {
  for(var key in config) {
    let variable = script.variables[key];
    if(variable) {
      variable.value = config[key];
    } else {
      throw new Error("Undeclared variable specified in configuration: " + key);
    }
  }
  verifyScript(script);
  script.background.forEach(function(block, i) {
    setTimeout(runBlock, 0, block, script, console.log.bind(console, "Background job", i, "terminated."));
  });
  script.listeners.forEach(function(listener, i) {
    switch(listener.protocol) {
    case 'http':
      if(typeof(listener.port) !== 'number') {
        throw new Error("Missing port number for http listener!");
      }
      require('http').createServer(function(req, res) {
        switch(req.method) {
        case 'OPTIONS':
          res.writeHead(204, CORS_HEADERS);
          res.end();
          break;
        case 'POST':
          {
            let args = req.url.slice(1).split('/'), action = args.shift();
            let actionBlock = script.actions[action];
            if(actionBlock) {
              runBlock(actionBlock, script, function(stack) {
                res.writeHead(200, {
                  'Content-Type': 'application/json',
                  'Access-Control-Allow-Origin': '*',
                  'Access-Control-Allow-Methods': 'POST'
                });
                res.write(JSON.stringify(stack, null, 2) + "\n");
                res.end();
              }, args);
            } else {
              res.writeHead(500, CORS_HEADERS);
              res.write("Unknown action: " + action + "\n");
              res.write("Have: " + JSON.stringify(Object.keys(script.actions)) + "\n");
              res.end();
            }
          }
          break;
        default:
          res.writeHead(400, CORS_HEADERS);
          res.write("Bad method!\n");
          res.end();
        }
      }).listen(listener.port);
      break;
    default:
      throw new Error("Unknown protocol: " + listener.protocol);
    }
  });
}

function runBlock(block, script, done, args) {
  // program counter
  let callIndex = 0;
  // stack management
  let stack = [];
  let PUSH = function(value) {
    if(typeof(value) === 'undefined') {
      throw new Error("PUSHED undefined!");
    }
    console.log('PUSH', typeof(value));
    stack.push(value);
  };
  let POP = function() {
    let value = stack.pop();
    if(typeof(value) === 'undefined') {
      throw new Error("Stack empty!");
    }
    console.log('POP', typeof(value));
    return value;
  };
  let POPLIST = function() {
    let value = POP();
    return (value instanceof Array) ? value : [value];
  };
  // push block arguments (as passed to actions)
  if(args) {
    args.forEach(PUSH);
  }
  if(block.timer) {
    // schedule block to be run
    if(typeof(block.timer.interval) !== 'undefined') {
      console.log('run every', block.timer.interval);
      setInterval(runCall, refValue(block.timer.interval, block));
      runCall();
    } else {
      setTimeout(runCall, refValue(block.timer.interval, block, 0));
    }
  } else {
    // run block now
    runCall();
  }
  function runCall() {
    if(callIndex === block.calls.length) {
      done(stack);
      return;
    }
    let call = block.calls[callIndex++];
    let next = setTimeout.bind(global, runCall, 0);
    if(! call) {
      next();
      return;
    }
    console.log('NEXT', '->', call.name);
    for(let a = (call.args.length - 1); a >= 0; a--) {
      PUSH(refValue(call.args[a], block));
    }
    switch(call.name) {
    case 'run-task':
      {
        let taskName = POP();
        let task = script.tasks[taskName];
        if(task) {
          console.log('RUN TASK', taskName);
          runBlock(task, script, next);
        } else {
          throw new Error("Task not found: " + task);
        }
      }
      break;
    case 'fetch':
      {
        let sources = POPLIST();
        console.log('FETCH SOURCES', sources);
        let accumulator = function(result) {
          this.push(result);
          if(this.length == sources.length) {
            PUSH(this);
            next();
          }
        }.bind([]);
        sources.forEach(function(source) {
          source.fetch(accumulator);
        });
      }
      break;
    case 'extract-feed-entries':
      {
        console.log('EXTRACT FEED ENTRIES');
        PUSH(POP().reduce(function(entries, feed) {
          return entries.concat(feed.entries.map(function(entry) {
            entry._id = entry.guid;
            return entry;
          }));
        }, []));
        next();
      }
      break;
    case 'set-flag':
      {
        console.log('SET FLAG');
        let key = POP(); //refValue(call.args[0], block);
        let value = POP(); //refValue(call.args[1], block);
        PUSH(POPLIST().map(function(record) {
          record[key] = value;
          return record;
        }));
        next();
      }
      break;
    case 'persist':
      {
        console.log('PERSIST');
        let datastore = POP(), records = POPLIST();
        let accumulator = function(record) {
          this.push(record);
          if(this.length === records.length) {
            PUSH(this);
            next();
          }
        }.bind([]);
        records.forEach(function(record) {
          if(! record._id) {
            throw new Error("Record is missing '_id': " + JSON.stringify(record));
          }
          datastore.store(record._id, record, accumulator.bind(this, record));
        });
      }
      break;
    case 'load':
      {
        console.log('LOAD');
        let datastore = POP();
        datastore.load(function(records) {
          console.log('loaded', records.length, 'records');
          PUSH(records);
          next();
        });
      }
      break;
    case 'load-item':
      {
        console.log('LOAD ITEM');
        let datastore = POP();
        let key = POP();
        if(typeof(key) !== 'string') {
          console.log('key', key);
          throw new Error("Invalid key :" + key + "!");
        }
        datastore.loadItem(key, function(record) {
          PUSH(record);
          next();
        });
      }
      break;
    case 'sort-by':
      {
        let field = POP();
        let direction = POP();
        let sorter = (
          direction === 'asc'
            ? function(_a, _b) {
              let a = _a[field], b = _b[field];
              if(typeof(a) == 'undefined' || typeof(b) == 'undefined') {
                return 0;
              }
              return a > b ? -1 : a < b ? 1 : 0;
            }
          : function(_a, _b) {
            let a = _a[field], b = _b[field];
            if(typeof(a) == 'undefined' || typeof(b) == 'undefined') {
              return 0;
            }
            return a < b ? -1 : a > b ? 1 : 0;
          }
        );
        PUSH(POPLIST().sort(sorter));
        next();
      }
      break;
    case 'filter':
      {
        let field = POP();
        let value = POP();
        PUSH(POPLIST().filter(function(item) {
          return item[field] === value;
        }));
        next();
      }
      break;
    default:
      console.log('STUB STATEMENT', call.name, call.args);
      next();
    }
  }
}

function verifyScript(script) {
  for(let name in script.tasks) {
    verifyBlock(script, script.tasks[name], "task", name);
  }
  script.background.forEach(function(block, i) {
    verifyBlock(script, block, " background", "#" + (i + 1));
  });
  for(let name in script.actions) {
    verifyBlock(script, script.actions[name], "action", name);
  }

  function verifyBlock(script, block, blockDesc, name) {
    block.references.forEach(function(reference) {
      if(reference.source && reference.source.required && (! reference.source.value)) {
        throw new Error("Variable not set: " + reference.symbol + " (required by: " + blockType + " " + name + ")");
      }
    });
  }
}

function findRef(symbol, block) {
  for(let i = 0, n = block.references.length; i < n; i++) {
    if(block.references[i].symbol === symbol) {
      return block.references[i];
    }
  }
  throw new Error("Reference not found in block: " + symbol);
}

function refValue(valueOrToken, block, defaultValue) {
  if(typeof(valueOrToken) === 'object' && valueOrToken.type === 'symbol') {
    let ref = findRef(valueOrToken.value, block);
    if(ref.source) {
      if(typeof(ref.source.value) === 'undefined') {
        if(typeof(defaultValue) === 'undefined') {
          throw new Error("Reference value not set: " + ref.symbol);
        } else {
          return defaultValue; // referenced value not found, but default given
        }
      } else {
        return ref.source.value; // referenced value found (such as with builtin)
      }
    }
  } else {
    return valueOrToken; // value wasn't a reference
  }
}

function evalScript(contexts) {
  let script = {
    variables: {},
    tasks: {},
    actions: {},
    background: [],
    listeners: []
  };
  contexts.forEach(function(context) {
    switch(context.type) {
    case 'define':
      script.variables[context.name] = evalDefinition(context.statements, {
        type: { type: 'symbol' },
        required: { type: 'boolean', emptyValue: true }
      });
      break;
    case 'listen':
      script.listeners.push(evalDefinition(context.statements, {
        protocol: { type: 'symbol' },
        address: { type: 'string' },
        port: { type: 'integer' }
      }));
      break;
    case 'task':
      script.tasks[context.name] = evalBlock(context.statements);
      break;
    case 'action':
      script.actions[context.name] = evalBlock(context.statements);
      break;
    case 'background':
      script.background.push(evalBlock(context.statements));
      break;
    default:
      throw new Error("Invalid context type: " + context.type);
    }
  });
  return script;
}

function evalDefinition(statements, keywords) {
  let definition = {};
  statements.forEach(function(statement) {
    let first = statement[0], second = statement[1];
    expectTokenType('symbol', first);
    let keyword = keywords[first.value];
    if(keyword) {
      if(keyword.emptyValue && (! second)) {
        definition[first.value] = keyword.emptyValue;
      } else {
        if(keyword.type) {
          expectTokenType(keyword.type, second);
        }
        definition[first.value] = second.value;
      }
    } else {
      throw new Error("Unknown statement keyword for definition: " + first.value);
    }
  });
  return definition;
}

function evalBlock(statements) {
  let block = {
    references: [],
    calls: []
  };
  statements.forEach(function(statement, statementIndex) {
    let first = statement[0];
    expectTokenType('symbol', first);
    block.references.push({
      type: 'call',
      symbol: first.value,
      statement: statementIndex
    });
    block.calls.push({
      name: first.value,
      args: statement.slice(1).map(function(arg, argIndex) {
        if(arg.type === 'symbol') {
          block.references.push({
            type: 'value',
            symbol: arg.value,
            statement: statementIndex,
            arg: argIndex
          });
          return arg;
        } else {
          return arg.value;
        }
      })
    });
  });
  return block;
}

function expectTokenType(type, token) {
  if(token.type !== type) {
    throw new Error("Unexpected token '" + token.type + "', expected '" + type + "'!");
  }
}

function resolveSymbols(script) {
  for(let name in script.tasks) {
    resolveBlockSymbols(script, script.tasks[name], "task", name);
  }
  script.background.forEach(function(block, i) {
    resolveBlockSymbols(script, block, " background", "#" + (i + 1));
  });
  for(let name in script.actions) {
    resolveBlockSymbols(script, script.actions[name], "action", name);
  }
}

function resolveBlockSymbols(script, block, blockType, name) {
  function expectArgs(ref, n, types) {
    if(ref.call.args.length !== n) {
      throw new Error("Expected " + n + " arguments for statement " + (ref.statement + 1) + ", but got " + ref.call.args.length + " instead! (from: " + blockType + " " + name + ")");
    }
    if(types) {
      types.forEach(function(type, i) {
        if(typeof(ref.call.args[i]) !== type) {
          throw new Error("Expected argument #" + (i+1) + " to be of type " + type + ", but got " + typeof(ref.call.args[i]) + " instead in statement " + (ref.statement + 1)+ " (from: " + blockType + " " + name + ")");
        }
      });
    }
  }
  block.references.forEach(function(reference) {
    if(reference.type === 'call') {
      console.log('resolve call  ', reference.symbol);
      reference.call = block.calls[reference.statement];
      switch(reference.symbol) {
      case 'every':
        expectArgs(reference, 1);
        if(block.timer) {
          throw new Error("Duplicate timer in block: " + blockType + " " + name);
        }
        block.timer = { interval: reference.call.args[0] };
        block.calls[reference.statement] = null;
        break;
      case 'run-task':
        expectArgs(reference, 1, ['string']);
        break;
      case 'persist':
        expectArgs(reference, 1);
        break;
      case 'fetch':
        expectArgs(reference, 1);
        break;
      case 'set-flag':
        expectArgs(reference, 2, ['string']);
        break;
      case 'extract-feed-entries':
        expectArgs(reference, 0);
        break;
      default:
        console.log('ERROR: unknown call', reference.symbol);
      }
    } else if(reference.type === 'value') {
      reference.source = ( script.variables[reference.symbol] ||
                           BUILTINS[reference.symbol]);
      if(! reference.source) {
        throw new Error("Failed to resolve symbol \"" + reference.symbol + "\" (from: " + blockType + " " + name + ", statement #" + (reference.statement + 1) + (typeof(reference.arg) !== 'undefined' ? ", argument " + (reference.arg + 1) : "") + ")");
      }
      console.log('resolve symbol ' + reference.symbol + ' -> ' + JSON.stringify(reference.source));
    } else {
      throw new Error("BUG: invalid reference type: " + reference.type);
    }
  });
}
